src/bin/pg_autoctl/state.c

```diff
@@ -492,29 +494,32 @@ NodeStateFromString(const char *str)
 /*
  * epoch_to_string converts a number of seconds from epoch into a date time
  * string.
+ *
+ * This string is stored in buffer and for ease of use of the function this
+ * buffer is also returned. The buffer should (at least) be MAXCTIMESIZE large.
  */
 const char *
-epoch_to_string(uint64_t seconds)
+epoch_to_string(uint64_t seconds, char *buffer)
 {
-	if (seconds > 0)
+	if (seconds <= 0) {
+		strlcpy(buffer, "0", MAXCTIMESIZE);
+		return buffer;
+	}
+	/*
+	 * ctime_r returns a string that ends with \n, which we don't want.
+	 * Replace it with a null string terminator, even if that means we lose
+	 * a precious byte of memory here..
+	 */
+	ctime_r((time_t *) &seconds, buffer);
+
+	if (buffer == NULL)
 	{
-		/*
-		 * ctime() returns a string that ends with \n, which we don't want.
-		 * Replace it with a null string terminator, even if that means we lose
-		 * a precious byte of memory here..
-		 */
-		char *str = ctime((time_t *) &seconds);
-
-		if (str == NULL)
-		{
-			log_error("Failed to convert epoch %" PRIu64 " to string: %m",
-					  seconds);
-			return NULL;
-		}
-		str[strlen(str) - 1] = '\0';
-		return str;
+		log_error("Failed to convert epoch %" PRIu64 " to string: %m",
+				  seconds);
+		return NULL;
 	}
-	return "0";
+	buffer[strlen(buffer) - 1] = '\0';
```

Comment:
Can the buffer be an empty string or something else that would trip us here? It looks suspicious the way we address the memory here.

Reply:
I now see that we used to do it that way already, so I guess it's fine. Still looks suspicious... but it does not belong to the current PR to solve it I guess.

Reply:
Added some sanity checks there.